### PR TITLE
New with spread

### PR DESCRIPTION
**Pull request by tinganho**
Implements #2369. Adds ES5 support for spread operator in `new expressions`. ES3 targets should still output warning as before, because it doesn't support `Function.prototype.bind` which I use in this ES5 implementation. I refactored some of the code from previous `call with spread implementation` #1931 so I could reuse it to on my implementation.

I have also manually tested the permutations of the outputted JavaScript.

### CODE REVIEW

**Comment by tinganho** [comment]
I had some serious trouble with `crlf` and `lf` line endings.

I have to turn off autocrlf:
```
git config --global core.autocrlf=false
```
In order for my test to pass. Have there been any experiencing the same problem as me? using Mac?

**Comment by CyrusNajmabadi** [comment]
Yes.  Don't use autocrlf :)   It's not appropriate for our project because line endings *are* an important thing we track and use in our tests.

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,7 @@ tests/services/baselines/local/*
 tests/baselines/prototyping/local/*
 tests/baselines/rwc/*
 tests/baselines/test262/*
+tests/baselines/reference/projectOutput/*
```
Comment (CyrusNajmabadi) [inline .gitignore]: why was this added to the .gitignore?

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -1880,14 +1859,73 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
+                //
+                //         var arguments = [1, 2, 3, 4, 5];
+                //         new (Array.bind.apply(Array, [null].concat(arguments)));
+                //
+                if (languageVersion === ScriptTarget.ES5 &&
+                    node.arguments &&
+                    hasSpreadElement(node.arguments)) {
+
+                    write('(');
+                    let { target, accessor } = emitCallOrNewExpressionTarget(node.expression);
```
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: i'm a little concerned that this approach will cause multiple evaluations of both the target and accessor.  Once in the code that is emitted in emitCallOrNewExpressionTarget and again in the code below.

File: .gitignore (hunk 1 above)
Comment (tinganho) [inline .gitignore]: I got multiple files outputted there every time I run `jake runtests`:

screen shot 2015-05-07 at 9 49 45 pm

File: src/compiler/emitter.ts (hunk 2 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: I'm not sure I follow. But it is suppose to emit the `target` every time and then emit the target and accessor again at below. Do you mean by evaluation that I don't store the string in the return value of `emitCallOrNewExpressionTarget()` instead storing Expression objects?

File: .gitignore (hunk 1 above)
Comment (mhegazy) [inline .gitignore]: @CyrusNajmabadi i think this is a fair one to add, ppl have made this mistake multiple times already. mainly because of a mocha issue on node 0.10.*

File: src/compiler/emitter.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/emitter.ts]: if the target is an expression, you are evaluating the expression multiple times at runtime..

```ts
function f() {
    if (random) return A;
    else return B;
}

new (f())(...args);
```
this now emits as:
```ts
new ((f()).bind.apply((f()), [null].concat(args))); // f() is evaluated twice.
```
where it should have been something like:
```ts
_f = f(), new (_f.bind.apply(_f, [null].concat(args))); 
var _f;
```

Comment (tinganho) [inline src/compiler/emitter.ts]: If I haven't missed something I'm pretty sure I got it covered(in my tests) it emits:
```typescript
new (((_f = f()).bind.apply(_f, [null].concat(args))); 
var _f;
```

Comment (mhegazy) [inline src/compiler/emitter.ts]: yes. my bad. sorry about that.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -1880,14 +1859,73 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
-                    write("(");
-                    emitCommaList(node.arguments);
-                    write(")");
+
+                // Spread operator logic can be supported in new expressions in ES5 using a combination
+                // of Function.prototype.bind() and Function.prototype.apply().
+                //
+                //     Example:
+                //
+                //         var arguments = [1, 2, 3, 4, 5];
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: The example is not complete:

```ts
var argumets = [1, 2, 3, 4, 5];
new Array(...arguments);
```

**Comment by mhegazy** [comment]
:+1: 
@vladima  and @JsonFreeman can you take a look.

File: src/compiler/emitter.ts (hunk 3 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: Ok I can add an example of that.

**Comment by tinganho** [comment]
I probably should re-style my tests to use `foo`, `bar`, `baz`, `x`, `y`, `z` and `A`, `B`, `C`.

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -1881,13 +1860,77 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
                 }
             }
 
+            function emitCallOrNewExpressionTarget(node: Expression): CallTarget {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: I think it is a little subtle that this function emits one thing (the target), and then returns two things, one of which it emitted and one of which it did not. Can you make it just return the target and the accessor, and then the caller will emit the target?

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -1881,13 +1860,77 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
+                //
+                if (languageVersion === ScriptTarget.ES5 &&
+                    node.arguments &&
+                    hasSpreadElement(node.arguments)) {
+
+                    write('(');
+                    let { target, accessor } = emitCallOrNewExpressionTarget(node.expression);
+                    write('.bind.apply(');
+                    if(target) {
+                        emit(target);
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Can you add an example of this case as well? Also, make sure this does not crash if you encounter something like `new super(...)`, which does not have an accessor.

File: src/compiler/emitter.ts (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: +1.  It should be named properly as well to convey that it's just computing hte call target, and not actually emitting anything.

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -1881,13 +1860,77 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
+                        write('.');
+                        emit(accessor);
+                    }
+                    else {
+                        emit(node.expression);
+                    }
+                    write(', [null].concat(');
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: This is not what we do for call expressions in ES5. See the comment at the top of the loop in emitListWithSpread. In your example, it would just be emitted as `new (Array.bind.apply(Array, arguments))`. If it had another argument, it would be `new (Array.bind.apply(Array, arguments.concat([otherArgument])))`.

File: .gitignore (hunk 1 above)
Comment (danquirk) [inline .gitignore]: yeah I don't think there's any real harm in having it

File: src/compiler/emitter.ts (hunk 4 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: @JsonFreeman sound like a good idea. I will address this later.

File: src/compiler/emitter.ts (hunk 6 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: Im not sure a follow. In the comments above I state that it will be emitted as`new (Array.bind.apply(Array, [null].concat(arguments)))` all the time. Just `new (Array.bind.apply(Array, arguments))` won't work, because the first in the argument list will be provided as `thisArg` to the constructor instead of being provided as the first argument in the constructor.

Sources:
[MDN: "this"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this)
[MDN: Function.prototype.bind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind)

File: src/compiler/emitter.ts (hunk 4 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: Just tried to address this. The reason I did it was so that I don't need to have multiple node kind checks — one for the getter and one for the emitter. We already have in `emitCallOrNewExpressionTarget()` the node kind checks:
```typescript
if (node.kind === SyntaxKind.PropertyAccessExpression) {
    ...
}
else if (node.kind === SyntaxKind.ElementAccessExpression) {
    ...
}
else if (node.kind === SyntaxKind.SuperKeyword) {
    ...
}
else {
    ...
}
```
If we would have one for getter and emitter the same node kind checks need to be applied mulitple times.

I don't have any perfect solution to this. Any idea how to fix this?

Comment (tinganho) [inline src/compiler/emitter.ts]: I just found a better way to do this. We can just call it `emitCallTarget()` and return the target and have a separate function for the emit of accessor or just inline the emit logic in `emitNewExpression` since we only use it there. 

The problem is though, that it already exists one function with `emitCallTarget()`. Any chance I can just rename it to `emitCallTargetWithOptionalTempAssignment()`?

```typescript
function emitCallTarget(node: Expression): Expression {
    if (node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.ThisKeyword || node.kind === SyntaxKind.SuperKeyword) {
        emit(node);
        return node;
    }
    let temp = createAndRecordTempVariable(TempFlags.Auto);

    write("(");
    emit(temp);
    write(" = ");
    emit(node);
    write(")");
    return temp;
}
```

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Well the functionality below looks a lot like `emitCallWithSpread`, which calls the `emitCallTarget` you've mentioned above, and you want to now call this `emitCallTarget`, so we should find some unifying scheme.

Also, on the issue of subtlety in behavior - while the name is important, I'd rather see documentation, which we are currently lacking in. Regardless of what this is called, please add a nice JSDoc comment to point out its peculiar behavior.

File: src/compiler/emitter.ts (hunk 5 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: I will look into it.

Comment (tinganho) [inline src/compiler/emitter.ts]: I got the one with no accessor already covered in my tests 
```typescript
let a: string[];
new f(1, 2, a...);
```
produces:
```typescript
var a;
new (f.bind.apply(f, [null].concat([1, 2].concat(a))));
```
I couldn't write a super call test without any warning and errors.

File: src/compiler/emitter.ts (hunk 4 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: I just tested my latest idea. It also requires additional node checks. `emitCallWithSpread()` also emits the accessor and writing the extra node checks is not worth it IMO.  Can't we just rename it to `emitCallTargetAndAccessor()`? It optionally emits and optionally returns the target and accessor. It is the best in term of DRY and also number of operations/comparison for the job.

File: src/compiler/emitter.ts (hunk 5 above)
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: You can add a super test with errors. The test is to make sure it doesn't crash.

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: I mean can you add an example **with** an accessor inline as a comment

File: src/compiler/emitter.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: @tinganho I don't think we're opposed to it, but let's see what the change looks like; it's hard to get a feel of it without seeing the end result. Worst-case scenario, we'll just revert the commits =)

File: src/compiler/emitter.ts (hunk 6 above)
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Oh I see what you mean. Because we are chaining two functions that take a thisArg, the first two arguments participate in the thisArg. That is very subtle. But in that case, I would expect it to be `new (Array.bind.apply(Array, [Array].concat(arguments)))`. Why is one of the thisArg instances Array and the other null?

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Looking at the source you indicated, it says
>When a bound function is used to construct a value, the provided this is ignored.
To me that means you can pass Array instead of null.

Comment (tinganho) [inline src/compiler/emitter.ts]: I'm not 100% sure on this. `thisArg` sets to the returning object of the constructor when we instantate it with `new`. Even though we bind it. It doesn't affect the `thisArg` when called with `new`. Therefore it is more effective to just use null then Array, since it doesn't matter.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Well I guess since it could be ignored, you could also pass nothing. And that also signifies the intent that it will be ignored. But I believe our way of passing nothing is `void 0`, rather than null. I think it does not matter. Any of the three will do.

Can you please augment your comment above to indicate that this value is the thisArg of bind, and will be ignored by the new operator? Thanks.

**Comment by tinganho** [comment]
#### Revision 2
I've just updated the function name to `emitCallTargetAndAccessor` to reflect the accessor too. It optionally emits and optionally returns the call target and accessor. I also added some comments about some very specific cases with the implementation. This also addresses a bug I found.

Bug1 (accessor is a string literal)
```typescript
new object["a-b"](1, 2, a..)

new (a_ = object)["a-b"].bind.apply(a_."a-b", [null]); // should be a["a-b"];
```

Bug2 (accessor is an numeric literal)
```typescript
new object["a-b"](1, 2, a..)

new (a_ = object)[1].bind.apply(a_.1, [null]); // should be a[1];
```

Old comments:
[.gitignore](https://github.com/Microsoft/TypeScript/pull/3066/files#r29852869)
[emitter.ts](https://github.com/Microsoft/TypeScript/pull/3066/files#r29961679)
[.types.ts](https://github.com/tinganho/TypeScript/commit/5cb0013180cb34af353c72244b800bf79f5b633f#commitcomment-11111785)

**Comment by tinganho** [comment]
@JsonFreeman I addressed the problems in the [old revision](https://github.com/tinganho/TypeScript/commit/8ca1a24d61315b0e60cd97e8ad4c8567311e92bf#commitcomment-11166963).

I only use `Function` now and it works. 

Also there was no wrong with the new expression call syntax `new object.Array()()`. But I added a test for it just in case.

Travis is pulling in a wrong commit that's why it is failing. But this one has the right commit https://travis-ci.org/tinganho/TypeScript/builds/62332258. (I think I can only fix this with a new PR though) I contacted travis to address this.

**Comment by JsonFreeman** [comment]
@tinganho Thanks for iterating on this implementation so well. One thing that would be helpful in the future is when you change something, push the changes as new commits instead of amending your old commits. That way it is very easy to see what changed over the course of the discussion.

File: tests/baselines/reference/newWithSpreadES5.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+var h;
+var i;
+// Basic expression
+new f(1, 2, "string");
+new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a))));
+new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a, ["string"]))));
+// Call expression
+new f(1, 2, "string")();
+new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a))))();
+new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a, ["string"]))))();
```
Comment (JsonFreeman) [inline tests/baselines/reference/newWithSpreadES5.js]: I do think that these two lines are incorrect because these really translate to
```ts
new f(1, 2, ...a);
```
and not
```ts
new f(1, 2, ...a)();
```

**Comment by JsonFreeman** [comment]
The one caveat with using Function as I suggested is that there could be a local declaration called Function that shadows the global function. Maybe the Function idea won't work then :disappointed:. Unless we emit some sort of capture, which is overkill I think. Do you think we should go back to the temp idea? Sorry to be fickle.

**Comment by tinganho** [comment]
>Do you think we should go back to the temp idea? Sorry to be fickle.

No problem, Ok I can change to your other idea with using temp for "target+accessor".

I will push them as new commits instead of cleaned ones.

File: tests/baselines/reference/newWithSpreadES5.js (hunk 7 above)
Comment (tinganho) [inline tests/baselines/reference/newWithSpreadES5.js]: I don't think I understand here can you elaborate? They have `()` at the end? Or does the outer parentheses `(Function...)` messing with it?

Comment (JsonFreeman) [inline tests/baselines/reference/newWithSpreadES5.js]: Yes, the outer parentheses are messing with it, because it causes the final `()` to be consumed as a part of the new expression. You need an extra `()` to make it a call expression. So you should emit:
```ts
new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a))))()();
```

**Comment by JsonFreeman** [comment]
Thanks. In terms of the implementation code, I guess my suggestion is that instead of using emitCallTargetAndAccessor, you just call emitCallTarget (the already existing one), and pass in node.expression as the argument (where `node` is the new expression).

**Comment by tinganho** [comment]
I think that is not so good idea because we want it to emit the accessor to?
For `new object.array`

it will set the temp to
```typescript
(a_ = object)
```

instead of:
```typescript
(a_ = object.array)
```

So my suggestion is to emit with a new `emitTempExpression` or something like that?

**Comment by tinganho** [comment]
hmm, I will try it out and see if there are any problem.

**Comment by JsonFreeman** [comment]
I think it should still give you `(a_ = object.array)` as the temp

**Comment by tinganho** [comment]
@JsonFreeman yes, I was little bit too fast with commenting without checking it out.

File: tests/baselines/reference/newWithSpreadES5.js (hunk 7 above)
Comment (tinganho) [inline tests/baselines/reference/newWithSpreadES5.js]: @JsonFreeman I guess if the call have arguments it should be compiled to:

```typescript
new (Function.bind.apply(f, [void 0].concat([1, 2].concat(a))))()(a, b);
```

**Comment by tinganho** [comment]
@JsonFreeman should we handle `new (...args)()` calls because the type checker only handles void functions right now?

Interestingly, I haven't got any error warnings about it when I pass arguments to a void function.

**Comment by tinganho** [comment]
I just pushed a new fix. (Excluding a fix on the type checker. And Travis is still pulling the wrong commit.)

**Comment by JsonFreeman** [comment]
I don't quite understand what you mean in this comment https://github.com/Microsoft/TypeScript/pull/3066#issuecomment-101898274

File: src/compiler/emitter.ts [hunk 8]
```diff
@@ -1850,6 +1850,12 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
                 emitListWithSpread(node.arguments, /*multiLine*/ false, /*trailingComma*/ false);
                 write(")");
             }
+            
+            function emitBracedEnclosedCommaList(nodes: Node[]) {
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: These are parens, not braces. But I don't think you need a separate function for this.

File: src/compiler/emitter.ts [hunk 9]
```diff
@@ -1875,19 +1888,68 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
-                    emitCommaList(node.arguments);
-                    write(")");
+                    let target = emitCallTarget(node.expression);
+                    write(".bind.apply(");
+                    emit(target);
+                    write(", [void 0].concat(");
+                    emitListWithSpread(node.arguments, /*multiline*/false, /*trailingComma*/false);
+                    write(")))");
+                    if (emitCallHelper) {
+                        write("()");
```
Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Oh, I see what you are doing. To simplify this even more, I would just always emit parentheses, regardless of whether the NewExpression is parented by a CallExpression. It's more idiomatic anyway.

Comment (tinganho) [inline src/compiler/emitter.ts]: I also thought about this at the beginning. I can fix this on a later commit.

**Comment by tinganho** [comment]
@JsonFreeman When you do:
```typescript
function f(...a): () => void {}
let a: string[];
new f(a...)()// error can only new a void function
```

But in my tests I only define the function as a void function:
```typescript
function f(...a){}
```

**Comment by JsonFreeman** [comment]
Right, if you define f as a void function, you cannot call it after you new it. But if you define it as a non-void function, you cannot new it in the first place.

It's okay that you get an error in these cases, the important thing here is that the emit is correct.

**Comment by tinganho** [comment]
Ok I just pushed a fix to always emit a pair of parens `()`

**Comment by JsonFreeman** [comment]
Beautiful!! Thanks for doing this, and great work.

So the only thing in the way now is Travis. It says there's a compile error. Are you seeing this locally as well?

**Comment by tinganho** [comment]
@JsonFreeman this is the same commit but on my repo https://travis-ci.org/tinganho/TypeScript/builds/62635340

Travis is pulling in the wrong commit.

**Comment by JsonFreeman** [comment]
I think Travis is saying that if your fork was merged into the main repo on master, it would fail. Can you try pulling the latest from upstream/master and see if that fixes it?

**Comment by tinganho** [comment]
@JsonFreeman it is still pulling in the wrong commit bde2491 instead of b88d542

**Comment by JsonFreeman** [comment]
Let's ask @DanielRosenwasser, since he knows more about Travis

**Comment by DanielRosenwasser** [comment]
@tinganho I pulled down your branch and I'm getting the same error:

> node bin/tsc.js --module commonjs -noImplicitAny --preserveConstEnums --out built\local\tsc.js -sourcemap -mapRoot file:///C:\Users\drosen\tinganho\built\local src\compiler\core.ts src\compiler\sys.ts src\compiler\types.ts src\compiler\scanner.ts src\compiler\parser.ts src\compiler\utilities.ts src\compiler\binder.ts src\compiler\checker.ts src\compiler\declarationEmitter.ts src\compiler\emitter.ts src\compiler\program.ts src\compiler\commandLineParser.ts src\compiler\tsc.ts src\compiler\diagnosticInformationMap.generated.ts

> src/compiler/emitter.ts(1916,21): error TS2346: Supplied parameters do not match any signature of call target.

This is related to some recent changes; you need to add the parameter `alwaysCopy` for `emitListWithSpread`.

Either way, you'll need to merge in recent changes.

File: src/compiler/emitter.ts [hunk 10]
```diff
@@ -1886,11 +1886,44 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
+                    hasSpreadElement(node.arguments)) {
+
                     write("(");
-                    emitCommaList(node.arguments);
-                    write(")");
+                    let target = emitCallTarget(node.expression);
+                    write(".bind.apply(");
+                    emit(target);
+                    write(", [void 0].concat(");
+                    emitListWithSpread(node.arguments, /*multiline*/false, /*trailingComma*/false);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Can you add spaces after each of these comments?

**Comment by tinganho** [comment]
@DanielRosenwasser I could get the errors now. Let me just fix it.

**Comment by JsonFreeman** [comment]
Oh wait, I noticed that
```ts
new f(...a, ...b);
```
will emit as
```js
new (f.bind.apply(f, [void 0].concat(a.concat(b)))();
```
And with alwaysCopy set to true, it would do
```ts
new f(...a)
```
as
```js
new (f.bind.apply(f, [void 0].concat(a.slice()))();
```
Can you confirm that this is what happens? This is not what we want, we don't want the slice or the second concat.

Can you add tests for this if it is not already affected?

File: src/compiler/emitter.ts [hunk 11]
```diff
@@ -1886,11 +1886,44 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
...
+                    hasSpreadElement(node.arguments)) {
+
                     write("(");
-                    emitCommaList(node.arguments);
-                    write(")");
+                    let target = emitCallTarget(node.expression);
+                    write(".bind.apply(");
+                    emit(target);
+                    write(", [void 0].concat(");
+                    emitListWithSpread(node.arguments, /*alwaysCopy*/true, /*multiline*/false, /*trailingComma*/false);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Can you actually rename `alwaysCopy` to `needsUniqueCopy` and replace the relevant comments elsewhere?

**Comment by JsonFreeman** [comment]
I think the reason the tests did not catch this is because all the tests have at least one non-spread argument before the spread.

**Comment by tinganho** [comment]
@JsonFreeman I was a little bit fast with the the push button. I think alwaysCopy should be set to false instead of true. Because it was false before.

I can add a test for it.

**Comment by JsonFreeman** [comment]
Sure, but you will still have the double concat problem.

File: src/compiler/emitter.ts (hunk 11 above)
Comment (tinganho) [inline src/compiler/emitter.ts]: Yes I can rename it.

**Comment by tinganho** [comment]
@JsonFreeman is that doable to do `f(...a, ...b)`? It throws an error now that a spread parameter can only be at the end of argument.

**Comment by JsonFreeman** [comment]
Regarding `f(...a, ...b)`. It is not legal to define a function with a rest parameter that is not at the end. So you can't define a function like this:
```ts
function f(...a, ...b) { }
```
But I meant if you have a function called f, and you call it with two spread arguments:
```ts
function f(...p: any[]) { }
var a: string[];
var b: string[];
f(...a, ...b);
```
And that is legal.

**Comment by tinganho** [comment]
@JsonFreeman I just tested this and it is outputting:
```typescript
new (f.bind.apply(f, [void 0].concat(a.concat(b)))();
```
I can't see what's wrong with it? What is the proposed emit?

**Comment by JsonFreeman** [comment]
That emit works, but I think it's nicer to emit:
```ts
new (f.bind.apply(f, [void 0].concat(a, b))();
```
You can achieve this by passing a boolean to emitListWithSpread to make it suppress the concat.

**Comment by tinganho** [comment]
Ok I fixed it on a new commit.

**Comment by JsonFreeman** [comment]
Sweet! Thanks @tinganho! I sign off. Please fix my one nit, and then I can merge it in.

**Comment by tinganho** [comment]
OK @JsonFreeman I fixed it on a new commit. Sorry for not seeing it earlier.

**Comment by JsonFreeman** [comment]
Thanks @tinganho!
